plutus-conformance/src/PlutusConformance/Common.hs

```diff
@@ -0,0 +1,134 @@
+import Test.Tasty.HUnit (testCase, (@=?))
+import Text.Megaparsec (SourcePos)
+import UntypedPlutusCore.Core.Type qualified as UPLC
+import UntypedPlutusCore.Evaluation.Machine.Cek (unsafeEvaluateCekNoEmit)
+import UntypedPlutusCore.Parser qualified as UPLC
+
+-- | A TestContent contains what you need to run a test.
+data TestContent =
+   MkTestContent {
+       -- | The path of the input file. This is also used to name the test.
+       testName    :: FilePath
+       -- | The expected output of the test in `Text`.
+       , expected  :: T.Text
+       -- | The input to be tested, in `Text`.
+       , inputProg :: T.Text
+   }
+
+{- | Create a list of `TestContent` with the given lists.
+The lengths of the input lists have to be the same, otherwise, an error occurs. -}
+mkTestContents ::
+    -- | The list of paths of the input files.
+    [FilePath] ->
+    -- | The list of expected outputs.
+    [T.Text] ->
+    -- | The list of the inputs.
+    [T.Text] ->
+    [TestContent]
+mkTestContents lFilepaths lRes lProgs =
+    case zipWith3Exact (\f r p -> MkTestContent f r p) lFilepaths lRes lProgs of
+        Just tests -> tests
+        Nothing -> error $ unlines
+            ["mkTestContents: Cannot run the tests because the number of input and output programs are not the same. "
+            , "Number of input files: "
+            , show (length lProgs)
+            , " Number of output files: "
+            , show (length lRes)
+            , " Make sure all your input programs have an accompanying .expected file."
+            ]
+
+{- | The default shown text when a parse error occurs.
+We don't want to show the detailed parse errors so that
+users of the test suite can produce this expected outputs more easily. -}
+shownParseError :: T.Text
+shownParseError = "parse error"
+
+-- | The default shown text when evaluation fails.
+shownEvaluationFailure :: T.Text
+shownEvaluationFailure = "evaluation failure"
+
+-- For UPLC evaluation tests
+
+type UplcProg = UPLC.Program Name DefaultUni DefaultFun ()
+
+termToProg :: UPLC.Term Name DefaultUni DefaultFun () -> UplcProg
+termToProg = UPLC.Program () (defaultVersion ())
+
+{- using the unsafe version of evaluate here so that it has a more generic signature.
+Any exceptions will be caught for any input runner in the tests, including our `evalUplcProg`.
+ -}
+evalUplcProg :: UplcProg -> EvaluationResult UplcProg
+evalUplcProg p =
+    fmap
+        termToProg
+        (unsafeEvaluateCekNoEmit defaultCekParameters (UPLC._progTerm p))
+
+{- | Run the inputs with the runner and return the results, in `Text`.
+When fail, the results are the default texts for parse error or evaluation failure. -}
+mkResult ::
+    -- | The `runner` to run the test inputs with.
+    (UplcProg -> EvaluationResult UplcProg) ->
+    -- | The result of parsing.
+    Either ParserErrorBundle (UPLC.Program Name DefaultUni DefaultFun SourcePos) ->
+    -- | The result in `Text`.
+        IO T.Text
```

Comment:
formatting nit, we usually put the arrows at the start of the line when we break type signatures across multiple lines